src/imputation/tmi_imputation.py

```diff
@@ -0,0 +1,337 @@
+
+    # Replace Nan with zero for companies with NO R&D Q604 = "No"
+    no_rd = filter_by_column_content(og_df, "604", ["No"])
+
+    for i in target_variables:
+        no_rd[i] = fill_zeros(no_rd, i)
+
+    og_df = apply_to_original(no_rd, og_df)
+
+    # Return cleaned original dataset
+    return og_df
+
+
+def tmi_pre_processing(df, target_variables_list: list) -> pd.DataFrame:
+
+    # Create a copy for reference
+    copy_df = df.copy()
+
+    # Filter for long form data
+    long_df = filter_by_column_content(copy_df, "formtype", ["0001"])
+
+    # Filter for good statuses
+    fil_df = apply_filters(long_df)
+
+    # Fill zeros for no r&d and apply to original
+
+    df = clean_no_rd_data(fil_df, df, target_variables_list)
+
+    # Calculate imputation classes for each row
+    imp_df = create_imp_class_col(df, "200", "201", "imp_class")
+
+    return imp_df
+
+
+def sort(target_variable: str, df: pd.DataFrame) -> pd.DataFrame:
+    sort_list = [
+        f"{target_variable}",
+        "employees",
+        "reference",
+    ]
+    sorted_df = df.sort_values(
+        by=sort_list,
+        ascending=[True, False, True],
+    )
+    # sorted_df.reset_index(drop=True, inplace=True)
+
+    return sorted_df
+
+
+def trim_check(
+    df: pd.DataFrame, check_value=10
+) -> pd.DataFrame:  # TODO add check_value to a cofig
+    """_summary_
+
+    Args:
+        df (pd.DataFrame, check_value, optional): _description_
+        Defaults to 10)->pd.DataFrame(.
+
+    Returns:
+        _type_: _description_
+    """
+    # tag for those classes with more than check_value (currently 10)
+    if len(df) <= check_value:  # TODO or is this just <
+        df["trim_check"] = "below_trim_threshold"
+    else:
+        df["trim_check"] = "above_trim_threshold"
+
+    return df
+
+
+def trim_bounds(
+    df: pd.DataFrame,
+    variable: str,
+    lower_perc=15,  # TODO add percentages to config -
+    # check method inBERD_imputation_spec_V3
+    upper_perc=15,
+) -> pd.DataFrame:
+    """_summary_
+
+    Args:
+        df (pd.DataFrame, lower_perc, optional): _description_.
+        Defaults to 15, TODO add percentages to config
+
+    Returns:
+        _type_: _description_
+    """
+    # trim only if more than 10
+    df["pre_index"] = df.index
+
+    if len(df) <= 10:
+        df[f"{variable}_trim"] = "dont trim"
+    else:
+        df = filter_by_column_content(df, "trim_check", ["above_trim_threshold"])
+        df.reset_index(drop=True, inplace=True)
+
+        # define the bounds for trimming
+        remove_lower = np.ceil(len(df) * (lower_perc / 100))
+        remove_upper = np.ceil(len(df) * (1 - upper_perc / 100))
+
+        # create trim tag (distinct from trim_check)
+        # to mark which to trim for mean growth ratio
+
+        df[f"{variable}_trim"] = "do trim"
+        df.loc[
+            remove_lower : remove_upper - 2, f"{variable}_trim"
+        ] = "dont trim"  # TODO check if needs to be inclusive of exlusive
```

Comment:
Please make this check before merging. How can this be checked? Who needs to be asked? Is there anything I can do to help?